### PR TITLE
Internalize reconnection mechanics and introduce connection pooling

### LINKED ISSUE #435: ApnsClient is unable to reconnect
After some time we get a ClientNotConnectedException and the ApnsClient is unable to reconnect. I tried several ways to to reconnect. First i implemented the suggestion like on the readme. 

```
if (e.getCause() instanceof ClientNotConnectedException) {
        System.out.println("Waiting for client to reconnect…");
        apnsClient.getReconnectionFuture().await();
        System.out.println("Reconnected.");
    }
```

After that was not working i tried to force a disconnect and connect again, but that's also not working:

```
apnsClient.disconnect().await(TIMEOUT);
logger.info("reconnect apns client after disconnection: " + apnsClient.isConnected());
apnsClient.connect(ApnsClient.PRODUCTION_APNS_HOST).await(TIMEOUT);
```
By the way the apnsClient.isConnected gives false, connect without exception but the next sending fails with ClientNotConenctedException. 

Currently only help is to restart the application.

Used Pushy 0.9.2 with netty-tcnative 1.1.33.Fork25

### LINKED ISSUE #454: Excessive memory usage after many reconnection attempts
I found some extra ordinary logs.

```
2017/04/20 01:24:45.722 DEBUG (Slf4JLogger.java:81) - SSL_read failed: OpenSSL error: error:10000414:SSL routines:OPENSSL_internal:SSLV3_ALERT_CERTIFICATE_REVOKED
2017/04/20 01:24:45.722  INFO (ApnsClient.java:523) - Failed to connect.
javax.net.ssl.SSLHandshakeException: error:10000414:SSL routines:OPENSSL_internal:SSLV3_ALERT_CERTIFICATE_REVOKED
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.shutdownWithError(ReferenceCountedOpenSslEngine.java:696)
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.sslReadErrorResult(ReferenceCountedOpenSslEngine.java:899)
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:853)
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:931)
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:974)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1097)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:968)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:902)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:411)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:373)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:351)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:373)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:129)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:651)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:574)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:488)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:450)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:873)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
	at java.lang.Thread.run(Thread.java:745)
2017/04/20 01:24:45.722  WARN (Slf4JLogger.java:146) - [id: 0x93247371, L:/10.0.13.69:41672 - R:api.push.apple.com/17.188.163.207:443] TLS handshake failed:
javax.net.ssl.SSLHandshakeException: error:10000414:SSL routines:OPENSSL_internal:SSLV3_ALERT_CERTIFICATE_REVOKED
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.shutdownWithError(ReferenceCountedOpenSslEngine.java:696)
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.sslReadErrorResult(ReferenceCountedOpenSslEngine.java:899)
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:853)
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:931)
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:974)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1097)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:968)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:902)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:411)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:373)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:351)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:373)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:129)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:651)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:574)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:488)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:450)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:873)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
	at java.lang.Thread.run(Thread.java:745)
2017/04/20 01:24:45.723 DEBUG (ApnsClient.java:471) - Disconnected. Next automatic reconnection attempt in 1 seconds.
```

----
 found the expired certificate. so that apnsClient tried to reconnect. 
In this case, when I connected at first time, the certificate was not expired yet. but after few days, it's been expired. 
how can I avoid this problem? I want to ignore to reconnect in this case.

### LINKED ISSUE #455: Internalize reconnection mechanics
Right now, we ask users to deal with temporarily-dropped connections by waiting on a reconnection future. This drives complexity internally and also for callers. The Pushy world would probably be a better place if we could hide these details internally.

One weakness with this approach is that callers could attempt to send a notification that will never succeed if the client never reconnects. As discussed in #444, though, it might make sense to add a per-notification timeout.

Clearly, we have some details to iron out here, but I think leaving our reconnection laundry out in the open is unnecessary.

### LINKED ISSUE #489: A certain period of time, network differential, Pushy constantly re connect APNs server, resulting in paralysis of the company network
image

Recently, frequently encountered bad network situation, Pushy constantly reconnection APNs server, a few seconds to produce more than 10000 connections. Has caused the company network are dragged down paralysis?

### PR DESCRIPTION
**Pull request by jchambers**
After several false starts, I'm pleased to share this work-in-progress attempt at channel pooling. In the past, the model was that a single `ApnsClient` would have a single connection to the APNs server. In this pull request, each client now has an `ApnsChannelPool` which manages access to one or more connections. Connections are created on demand, which eliminates the notion of explicitly connecting (or reconnecting) an `ApnsClient`.

Because the model is changing significantly, this change will introduce lots of breaking API changes. I'm very interested in all of your feedback.

Again, I emphasize that this is a work in progress, but I do think it should be fairly representative of the shape of things to come. I'm also hopeful that this will resolve the host of reconnection issues folks have been reporting lately. In particular, I believe this:

- Fixes #435 
- Fixes #454 
- Closes #455 
- Fixes #489 

TODO:

- [x] Restore/update metrics
- [x] Restore exponential backoff when opening new connections
- [x] Add factory/pool tests
- [x] Document everything

### CODE REVIEW

File: benchmark/src/main/java/com/turo/pushy/apns/ApnsClientBenchmark.java [hunk 1]
```diff
@@ -52,6 +52,9 @@
     @Param({"10000"})
     public int notificationCount;
 
+    @Param({"1", "4", "8"})
+    public int concurrentConnections;
```
Comment (jchambers) [inline benchmark/src/main/java/com/turo/pushy/apns/ApnsClientBenchmark.java]: For the curious, early benchmarks show things to be maaaaaaaaybe slightly faster (it's pretty much in the noise) than they were before channel pooling with a single connection. Unsurprisingly, a single client's throughput goes up significantly when using multiple threads.

**Comment by clslrns** [comment]
This one would be very useful.
I've managed to get only around 11K notifications/s from single connection with 8-threaded event loop group on machine with 8 cores. Neither network, nor CPU was fully loaded. So I switched to 8 single-threaded `APNSClient`s and throughput improved to 35-44K notifications/s.

**Comment by jchambers** [comment]
Aside from the part where I think the Travis CI team is trolling me by constantly fixing and then un-fixing a weird hostname bug for OpenJDK 7, I think this is ready to roll. I'll sleep on it before I merge it, but anticipate a release some time tomorrow.

**Comment by jchambers** [comment]
For those of you waiting patiently for this change (and release): given the significance of the changes, we want to observe it in production for a little while on our end before we unleash it upon the wider world. I'm hoping for a release at the end of next week.

**Comment by jchambers** [comment]
Our work in #515 prompted some testing with very high levels of leak detection, which in turn increased load and slowed everything down. That, in turn, revealed some sporadic test failures when sending many notifications, which hints at a race condition. We're tracking it down, but it's something we'll need to resolve before the next release.

**Comment by jchambers** [comment]
Here's what I've figured out so far:

- The "send many notifications" tests are sometimes failing because a "send" future fails because the stream was closed before the server sent a reply.
- The stream appears to have been closed because the underlying channel was closed.
- The underlying channel appears to have been closed because we tried to acquire it from the idle pool, but then noticed that it wasn't active.

The last part is really puzzling to me, and will be the focus of the investigation from here.

**Comment by jchambers** [comment]
In all cases of failure, the client sends a `GOAWAY` frame to the server, and the last stream created by the client is 143. This is shaping up to be either a spectacularly dumb or really interesting bug.

**Comment by jchambers** [comment]
Found it. We're attaching a response promise to streams in the client handler asynchronously, and in some cases, we can get a reply from the server before that promise is in place, which causes a `NullPointerException`. This was likely also a problem in 0.10, but we just didn't notice it until now.

**Comment by jchambers** [comment]
Everything is fixed in ed1d764.

**Pull request merged**